cleanrl/experiments/sac_online_qtargs_disc.py

```diff
@@ -0,0 +1,402 @@
+    parser.add_argument('--gamma', type=float, default=0.99,
+                       help='the discount factor gamma')
+    parser.add_argument('--target-update-interval', type=int, default=1,
+                       help="the timesteps it takes to update the target network")
+    parser.add_argument('--batch-size', type=int, default=64,
+                       help="the batch size of sample from the reply memory")
+    parser.add_argument('--tau', type=float, default=0.005,
+                       help="target smoothing coefficient (default: 0.005)")
+    parser.add_argument('--alpha', type=float, default=0.2,
+                       help="Entropy regularization coefficient.")
+    parser.add_argument('--autotune', action='store_true',
+        help='Enables autotuning of the alpha entropy coefficient')
+
+    # Neural Network Parametrization
+    parser.add_argument('--policy-hid-sizes', nargs='+', type=int, default=(120,84,))
+    parser.add_argument('--value-hid-sizes', nargs='+', type=int, default=(120,84,))
+    parser.add_argument('--q-hid-sizes', nargs='+', type=int, default=(120,84,))
+
+    args = parser.parse_args()
+    if not args.seed:
+        args.seed = int(time.time())
+
+# TRY NOT TO MODIFY: setup the environment
+device = torch.device('cuda' if torch.cuda.is_available() and args.cuda else 'cpu')
+env = gym.make(args.gym_id)
+random.seed(args.seed)
+np.random.seed(args.seed)
+torch.manual_seed(args.seed)
+torch.backends.cudnn.deterministic = args.torch_deterministic
+env.seed(args.seed)
+env.action_space.seed(args.seed)
+env.observation_space.seed(args.seed)
+input_shape, preprocess_obs_fn = preprocess_obs_space(env.observation_space, device)
+output_shape = preprocess_ac_space(env.action_space)
+assert isinstance(env.action_space, Discrete), "only discrete action space is supported"
+
+# ALGO LOGIC: initialize agent here:
+EPS = 1e-8
+LOG_STD_MAX = 2
+LOG_STD_MIN = -20
+
+# Custom Categorical Policy
+class Policy(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self._layers = nn.ModuleList()
+
+        current_dim = input_shape
+        for hsize in list( args.policy_hid_sizes) + list( [output_shape,]):
+            self._layers.append( nn.Linear( current_dim, hsize))
+            current_dim = hsize
+
+        self.nn_softmax = torch.nn.Softmax(1)
+        self.nn_log_softmax = torch.nn.LogSoftmax(1)
+
+    def forward(self, x):
+        # # Workaround Seg. Fault when passing tensor to Q Function
+        if not isinstance( x, torch.Tensor):
+            x = preprocess_obs_fn(x)
+
+        for layer in self._layers[:-1]:
+            x = F.relu( layer(x))
+
+        # Return the logits [batch_size, output_shape]
+        logits = self._layers[-1]( x)
+
+        return self.nn_softmax( logits), self.nn_log_softmax( logits)
+
+    # TODO: Since we use torch.nn.{Softmax, LogSoftmax}, get_action_probs loses
+    # most of it s meaning.
+    def get_action_probs( self, observations):
```

Comment:
I guess you prefer using forward() ?

Reply:
No, in some cases like vanilla a2c (https://github.com/vwxyzjn/cleanrl/blob/9f348d082f217b1605a9df4922ec88763cfa4aec/cleanrl/a2c.py#L130), if we addd a `get_action` function, it will only be called once. So I think it's not necessary to include it. However, in other cases like PPO (https://github.com/vwxyzjn/cleanrl/blob/9f348d082f217b1605a9df4922ec88763cfa4aec/cleanrl/ppo.py#L133, https://github.com/vwxyzjn/cleanrl/blob/9f348d082f217b1605a9df4922ec88763cfa4aec/cleanrl/ppo.py#L175) it makes more sense to have a `get_action` function.

Reply:
I am now inclined to include a `get_action` function as it will also be useful for training using replay buffer in general.